### PR TITLE
feat(#8229): schedule outbound push

### PR DESCRIPTION
**Pull request by samuelimoisili**
# Description

Added the ability schedule outbound push using cron expression.

medic/cht-core#8229

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8229-outbound-push/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8229-outbound-push/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:8229-outbound-push/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: sentinel/tests/unit/schedule/outbound.js
Comment (lorerod) [inline sentinel/tests/unit/schedule/outbound.js:751]: Thank you for adding this scenario to the outbound.js unit test.

File: shared-libs/transitions/test/unit/lib/utils.js
Comment (lorerod) [inline shared-libs/transitions/test/unit/lib/utils.js:297]: Great! Thanks for this!

File: shared-libs/transitions/src/transitions/mark_for_outbound.js
Comment (lorerod) [inline shared-libs/transitions/src/transitions/mark_for_outbound.js:31]: What do you think about adding a unit test for this file? We can test the conditions added with your changes.

**Review by lorerod — CHANGES_REQUESTED** [review]
Thank you, @samuelimoisili, for these changes. 
I just left some notes related to testing. 
We can also add scenarios to the integration tests in tests/integration/sentinel/transitions/mark-for-outbound.spec.js or tests/integration/sentinel/schedules/outbound.spec.js. 
Let me know what you think.

File: shared-libs/transitions/src/transitions/mark_for_outbound.js
Comment (samuelimoisili) [reply shared-libs/transitions/src/transitions/mark_for_outbound.js:31]: Yes, I think it's a great idea. I'll be adding a test file for `mark_for_outbound.js`.

**Comment by samuelimoisili** [comment]
> We are also going to need to update [this](https://docs.communityhealthtoolkit.org/apps/reference/app-settings/outbound/) docs to add the new "cron" property.

Yes, I'll create a PR as soon as this get merged on cht-docs.

File: shared-libs/transitions/src/lib/utils.js
Comment (dianabarsan) [inline shared-libs/transitions/src/lib/utils.js:263]: I'm trying to understand how this works. 
Suppose that my cron is `* 6 * * *` and the current date time is `6:49:20`.
What the code does right now: 
- computes a date which is today at 6:50:00
- computes a date which is today at 6:49:00
- we then do: `6:49:20` >= `6:50:00` || `6:49:20` <= ` 6:49:00` which obviously returns false

I believe, in this example, the function should return true, any time between `6:00` and `6:59:59` should return true.

File: shared-libs/transitions/test/unit/lib/utils.js
Comment (dianabarsan) [inline shared-libs/transitions/test/unit/lib/utils.js:298]: I believe more cron options should be tested, instead of just one. I would go for the whole array of fields (minute, hour, day, etc) and test with a cron for each of them.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work! I left a few comments inline.

File: shared-libs/transitions/src/lib/utils.js
Comment (samuelimoisili) [reply shared-libs/transitions/src/lib/utils.js:136]: `later` returns the current date as the next schedule and it doesn't have validation built in. Okay, I'll install a cron validation library.

File: shared-libs/transitions/src/lib/utils.js
Comment (samuelimoisili) [reply shared-libs/transitions/src/lib/utils.js:263]: Not exactly, here's how it's works:

```js
_current_time = '2023-01-01 02:00:30' // time sentinel fires

exp =  '0 2 * * *' // runs on 2am precisely
frame = 60000 // 1 minute

calculation = '2023-01-01 02:01:00' >= '2023-01-01 02:00:00' || '2023-01-01 02:01:00' <= '2023-02-01 02:01:00'

returns = true // current time is within 1 minute of the expected fire time.
```

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Haven't reviewed code yet. Just noticed that the tests are failing because API can't start:
```
2023-07-26 17:47:56 [31mERROR[39m: Fatal error initialising API 
2023-07-26 17:47:56 [31mERROR[39m: Error: Cannot find module 'later'
Require stack:
- /service/shared-libs/transitions/src/lib/utils.js
- /service/shared-libs/transitions/src/transitions/index.js
- /service/shared-libs/transitions/src/index.js
- /service/api/src/services/config-watcher.js
- /service/api/server.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1028:15)
    at Function.Module._load (node:internal/modules/cjs/loader:873:27)
    at Module.require (node:internal/modules/cjs/loader:1100:19)
    at require (node:internal/modules/cjs/helpers:108:18)
    at Object.<anonymous> (/service/shared-libs/transitions/src/lib/utils.js:10:15)
    at Module._compile (node:internal/modules/cjs/loader:1198:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
    at Module.load (node:internal/modules/cjs/loader:1076:32)
    at Function.Module._load (node:internal/modules/cjs/loader:911:12)
    at Module.require (node:internal/modules/cjs/loader:1100:19) {
  [stack]: "Error: Cannot find module 'later'\n" +
    'Require stack:\n' +
    '- /service/shared-libs/transitions/src/lib/utils.js\n' +
    '- /service/shared-libs/transitions/src/transitions/index.js\n' +
    '- /service/shared-libs/transitions/src/index.js\n' +
    '- /service/api/src/services/config-watcher.js\n' +
    '- /service/api/server.js\n' +
    '    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1028:15)\n' +
    '    at Function.Module._load (node:internal/modules/cjs/loader:873:27)\n' +
    '    at Module.require (node:internal/modules/cjs/loader:1100:19)\n' +
    '    at require (node:internal/modules/cjs/helpers:108:18)\n' +
    '    at Object.<anonymous> (/service/shared-libs/transitions/src/lib/utils.js:10:15)\n' +
    '    at Module._compile (node:internal/modules/cjs/loader:1198:14)\n' +
    '    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)\n' +
    '    at Module.load (node:internal/modules/cjs/loader:1076:32)\n' +
    '    at Function.Module._load (node:internal/modules/cjs/loader:911:12)\n' +
    '    at Module.require (node:internal/modules/cjs/loader:1100:19)',
  [message]: "Cannot find module 'later'\n" +
    'Require stack:\n' +
    '- /service/shared-libs/transitions/src/lib/utils.js\n' +
    '- /service/shared-libs/transitions/src/transitions/index.js\n' +
    '- /service/shared-libs/transitions/src/index.js\n' +
    '- /service/api/src/services/config-watcher.js\n' +
    '- /service/api/server.js',
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/service/shared-libs/transitions/src/lib/utils.js',
    '/service/shared-libs/transitions/src/transitions/index.js',
    '/service/shared-libs/transitions/src/index.js',
    '/service/api/src/services/config-watcher.js',
    '/service/api/server.js',
    [length]: 5
  ]
} 
```

File: api/package-lock.json
Comment (dianabarsan) [inline api/package-lock.json:18]: It's a little weird that this is needed here. Since we don't actually alter API code. I'm noticing that there is no update of `shared-libs/transitions/package-lock.json`, but dependencies were added to the package.json file.

File: shared-libs/transitions/test/unit/transitions/mark_for_outbound.js
Comment (dianabarsan) [inline shared-libs/transitions/test/unit/transitions/mark_for_outbound.js:30]: I think you can do this a simpler way like: 
```suggestion
require('path/to/config')
sinon.stub(config, 'get');
```
Same for db, logger, I believe. Using `rewire` is less readable and is not what we practice most times. I would only use rewire when strictly required. 
Also, keeping the `restores` property is so cumbersome. 
I would rather we redo the loading for every test instead of keeping a list of properties to restore - which is more prone to mistakes. 
```
let markForOutbound;
beforeEach(() => {
  markForOutbound = rewire('../../../src/transitions/mark_for_outbound');
});
```

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Nice work! I gave this a quick look and left some comments, especially about resolving and installing dependencies. 
Additionally, can you please add an e2e test for outbound tasks with crons?

File: api/package-lock.json
Comment (samuelimoisili) [reply api/package-lock.json:18]: I noticed there was a recent switch to workspaces. If npm workspaces is like yarn workspaces, there's only supposed to be one lockfile and that is in the root. Here's a link to [npm workspace doc](https://docs.npmjs.com/cli/v8/using-npm/workspaces#:~:text=Below%20is%20a%20post%20npm%20install%20example%2C%20given%20that%20same%20previous%20example%20structure%20of%20files%20and%20folders%3A) insinuating it, but it's not explicitly stated like in the yarn documentation. `api` and `sentinel` are not part of the workspace that's why their lockfiles are being respected.

File: package-lock.json
Comment (samuelimoisili) [reply package-lock.json:11958]: Yep, it is. Workspaces in NodeJS only respect one lockfile which is the root directory's lockfile.

Comment (dianabarsan) [reply package-lock.json:11958]: We only use workspaces for shared libs - for actually injecting the shared libs themselves into apps: webapp, api, sentinel. we don't use them for other dependencies. Please keep the dependencies in the scope where they are to be used.

File: api/package-lock.json
Comment (dianabarsan) [reply api/package-lock.json:18]: We only use workspaces for shared libs - for actually injecting the shared libs themselves into apps: webapp, api, sentinel. we do

File: shared-libs/transitions/test/unit/transitions/mark_for_outbound.js
Comment (lorerod) [inline shared-libs/transitions/test/unit/transitions/mark_for_outbound.js:1]: Thanks for adding this!

**Review by lorerod — CHANGES_REQUESTED** [review]
Thanks your @samuelimoisili for adding my suggestions. 
Please, don't forget to include end-to-end tests as recommended in [this comment](https://github.com/medic/cht-core/pull/8387#pullrequestreview-1528999119).

File: api/package-lock.json
Comment (samuelimoisili) [reply api/package-lock.json:18]: I've removed the packages, but CI keeps failing. If you have any ideas on how to fix it, please help me out. Thanks.

File: package-lock.json
Comment (samuelimoisili) [reply package-lock.json:11958]: I've removed the packages, but CI keeps failing. If you have any ideas on how to fix it, please help me out. Thanks.

File: shared-libs/transitions/package.json
Comment (samuelimoisili) [inline shared-libs/transitions/package.json:29]: @dianabarsan The package-lock.json isn't updated after the install, it keeps routing to the root package-lock.json.

Comment (dianabarsan) [reply shared-libs/transitions/package.json:29]: Uh, you're right! It's because of npm workspaces. I've been trying to find some way to get npm to install the package in the correct place, but no luck yet. We're going to have to find a solution for this. 

cc @garethbowen

Comment (garethbowen) [reply shared-libs/transitions/package.json:29]: What's the downside of putting them in the root?

Comment (dianabarsan) [reply shared-libs/transitions/package.json:29]: they don't get hoisted in the docker image :(

Comment (garethbowen) [reply shared-libs/transitions/package.json:29]: Damn, true.

We need to find a proper solution, but to unblock @samuelimoisili can we inline the dependency in the sentinel package.json?

Comment (dianabarsan) [reply shared-libs/transitions/package.json:29]: @samuelimoisili do you think you can try @garethbowen 's suggestion?

File: tests/integration/sentinel/schedules/outbound.spec.js
Comment (lorerod) [inline tests/integration/sentinel/schedules/outbound.spec.js:183]: Can you please delete the commented lines? And to unblock you I can create another issue to complete this scenario.

**Review by lorerod — APPROVED** [review]
Thanks @samuelimoisili for implementing mark for outbound integration test.

**Comment by njuguna-n** [comment]
@dianabarsan please have another look.

File: api/package.json
Comment (dianabarsan) [inline api/package.json:31]: I _think_ we only need these in the shared library. So both `later` and `cron-validator` can be removed from `api` as dependencies.

File: sentinel/package.json
Comment (dianabarsan) [inline sentinel/package.json:14]: As above, I _think_ it's sufficient to add this as a dependency in `shared-libs/transitions`, where it will end up being used.

File: tests/integration/sentinel/schedules/outbound.spec.js
Comment (dianabarsan) [inline tests/integration/sentinel/schedules/outbound.spec.js:183]: Can we handle these lines somehow?

File: tests/integration/sentinel/transitions/mark-for-outbound.spec.js
Comment (dianabarsan) [inline tests/integration/sentinel/transitions/mark-for-outbound.spec.js:192]: I believe `sinon.useFakeTimers` won't work here. You're changing time in the test, not on the server code that actually runs. As far as I can tell, integration tests aren't actually running - the suite has been active for 4 hours so something's definitely wrong. 
I think you need a different approach here - maybe dynamically determine the cron value depending on server time at the moment of test running.

**Review by dianabarsan — CHANGES_REQUESTED** [review]
Thanks for picking this up @njuguna-n . I left a couple of comments inline.

File: api/package.json
Comment (lorerod) [reply api/package.json:31]: I tried to remove `later `and `cron-validator` from `api/package.json` and `sentinel/package.json` but got this error in each case:

```
2023-09-28 09:47:59 2023-09-28T13:47:59.960 ERROR: Fatal error intialising sentinel 
2023-09-28 09:47:59 2023-09-28T13:47:59.961 ERROR: Error: Error loading configuration
2023-09-28 09:47:59     at /service/sentinel/src/config.js:75:13
2023-09-28 09:47:59     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
2023-09-28 09:47:59   [stack]: 'Error: Error loading configuration\n' +
2023-09-28 09:47:59     '    at /service/sentinel/src/config.js:75:13\n' +
2023-09-28 09:47:59     '    at processTicksAndRejections (node:internal/process/task_queues:96:5)',
2023-09-28 09:47:59   [message]: 'Error loading configuration'
2023-09-28 09:47:59 } 
2023-09-28 09:48:13 Info: Starting CHT Sentinel
2023-09-28 09:48:14 2023-09-28T13:48:14.437 INFO: Running server checks... 
2023-09-28 09:48:14 Node Environment Options: 'undefined'
2023-09-28 09:48:14 Node Version: 16.20.1 in development mode
2023-09-28 09:48:14 CouchDB Version: 3.3.2
2023-09-28 09:48:14 CouchDb Cluster ready
2023-09-28 09:48:14 2023-09-28T13:48:14.647 INFO: API is ready: {"ready":true,"handler":"medic-api","version":"0.1.0","detail":"All required services are running normally"} 
2023-09-28 09:48:15 2023-09-28T13:48:15.223 ERROR: Error: Cannot find module 'cron-validator'
2023-09-28 09:48:15 Require stack:
2023-09-28 09:48:15 - /service/shared-libs/transitions/src/lib/utils.js
2023-09-28 09:48:15 - /service/shared-libs/transitions/src/transitions/index.js
2023-09-28 09:48:15 - /service/shared-libs/transitions/src/index.js
2023-09-28 09:48:15 - /service/sentinel/src/config.js
2023-09-28 09:48:15 - /service/sentinel/server.js
2023-09-28 09:48:15     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1028:15)
2023-09-28 09:48:15     at Function.Module._load (node:internal/modules/cjs/loader:873:27)
2023-09-28 09:48:15     at Module.require (node:internal/modules/cjs/loader:1100:19)
2023-09-28 09:48:15     at require (node:internal/modules/cjs/helpers:108:18)
2023-09-28 09:48:15     at Object.<anonymous> (/service/shared-libs/transitions/src/lib/utils.js:3:14)
2023-09-28 09:48:15     at Module._compile (node:internal/modules/cjs/loader:1198:14)
2023-09-28 09:48:15     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)
2023-09-28 09:48:15     at Module.load (node:internal/modules/cjs/loader:1076:32)
2023-09-28 09:48:15     at Function.Module._load (node:internal/modules/cjs/loader:911:12)
2023-09-28 09:48:15     at Module.require (node:internal/modules/cjs/loader:1100:19) {
2023-09-28 09:48:15   [stack]: "Error: Cannot find module 'cron-validator'\n" +
2023-09-28 09:48:15     'Require stack:\n' +
2023-09-28 09:48:15     '- /service/shared-libs/transitions/src/lib/utils.js\n' +
2023-09-28 09:48:15     '- /service/shared-libs/transitions/src/transitions/index.js\n' +
2023-09-28 09:48:15     '- /service/shared-libs/transitions/src/index.js\n' +
2023-09-28 09:48:15     '- /service/sentinel/src/config.js\n' +
2023-09-28 09:48:15     '- /service/sentinel/server.js\n' +
2023-09-28 09:48:15     '    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1028:15)\n' +
2023-09-28 09:48:15     '    at Function.Module._load (node:internal/modules/cjs/loader:873:27)\n' +
2023-09-28 09:48:15     '    at Module.require (node:internal/modules/cjs/loader:1100:19)\n' +
2023-09-28 09:48:15     '    at require (node:internal/modules/cjs/helpers:108:18)\n' +
2023-09-28 09:48:15     '    at Object.<anonymous> (/service/shared-libs/transitions/src/lib/utils.js:3:14)\n' +
2023-09-28 09:48:15     '    at Module._compile (node:internal/modules/cjs/loader:1198:14)\n' +
2023-09-28 09:48:15     '    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1252:10)\n' +
2023-09-28 09:48:15     '    at Module.load (node:internal/modules/cjs/loader:1076:32)\n' +
2023-09-28 09:48:15     '    at Function.Module._load (node:internal/modules/cjs/loader:911:12)\n' +
2023-09-28 09:48:15     '    at Module.require (node:internal/modules/cjs/loader:1100:19)',
2023-09-28 09:48:15   [message]: "Cannot find module 'cron-validator'\n" +
2023-09-28 09:48:15     'Require stack:\n' +
2023-09-28 09:48:15     '- /service/shared-libs/transitions/src/lib/utils.js\n' +
2023-09-28 09:48:15     '- /service/shared-libs/transitions/src/transitions/index.js\n' +
2023-09-28 09:48:15     '- /service/shared-libs/transitions/src/index.js\n' +
2023-09-28 09:48:15     '- /service/sentinel/src/config.js\n' +
2023-09-28 09:48:15     '- /service/sentinel/server.js',
2023-09-28 09:48:15   code: 'MODULE_NOT_FOUND',
2023-09-28 09:48:15   requireStack: [
2023-09-28 09:48:15     '/service/shared-libs/transitions/src/lib/utils.js',
2023-09-28 09:48:15     '/service/shared-libs/transitions/src/transitions/index.js',
2023-09-28 09:48:15     '/service/shared-libs/transitions/src/index.js',
2023-09-28 09:48:15     '/service/sentinel/src/config.js',
2023-09-28 09:48:15     '/service/sentinel/server.js',
2023-09-28 09:48:15     [length]: 5
2023-09-28 09:48:15   ]
2023-09-28 09:48:15 } 
2023-09-28 09:48:15 2023-09-28T13:48:15.224 ERROR: Fatal error intialising sentinel 
2023-09-28 09:48:15 2023-09-28T13:48:15.228 ERROR: Error: Error loading configuration
2023-09-28 09:48:15     at /service/sentinel/src/config.js:75:13
2023-09-28 09:48:15     at processTicksAndRejections (node:internal/process/task_queues:96:5) {
2023-09-28 09:48:15   [stack]: 'Error: Error loading configuration\n' +
2023-09-28 09:48:15     '    at /service/sentinel/src/config.js:75:13\n' +
2023-09-28 09:48:15     '    at processTicksAndRejections (node:internal/process/task_queues:96:5)',
2023-09-28 09:48:15   [message]: 'Error loading configuration'
2023-09-28 09:48:15 } 
```

File: sentinel/package.json
Comment (lorerod) [reply sentinel/package.json:14]: Same case as [this](https://github.com/medic/cht-core/pull/8387#discussion_r1340196350) comment

**Comment by lorerod** [comment]
@njuguna-n @njogz @dianabarsan, please have another look. I tried to fix the integration test `tests/integration/sentinel/transitions/mark-for-outbound.spec.js` and `tests/integration/sentinel/schedules/outbound.spec.js`. 
Locally, the integration tests are passing, but in CI, I'm getting the following:

```
Error response from daemon: No such container: cht-e2e_sentinel_1
docker exec date failed. NOTE this error is not relevant if running outside of docker
Command failed: docker exec cht-e2e_sentinel_1 date '+%Y-%m-%d %H:%M:%S'
Error response from daemon: No such container: cht-e2e_sentinel_1
```
When getting the container date in line [tests/utils/index.js](https://github.com/medic/cht-core/blob/2c8ca98a52f86853b443541d7cabbe595bcaa44d/tests/utils/index.js#L1225)

Could you please provide me with some suggestions?
Thanks!

File: tests/integration/sentinel/schedules/outbound.spec.js
Comment (dianabarsan) [inline tests/integration/sentinel/schedules/outbound.spec.js:5]: I think the issue you have is that you are calling this function here, when the file is loaded, before containers are up. Can you try moving this in the beforeEach or before ?

File: sentinel/package.json
Comment (dianabarsan) [reply sentinel/package.json:14]: You're right! Apparently, with workspaces, you no longer store dependencies in workspaces :(

File: api/package.json
Comment (dianabarsan) [reply api/package.json:31]: :( workspaces are just weird

File: tests/utils/index.js
Comment (dianabarsan) [inline tests/utils/index.js:1228]: Please use `console.error` when logging errrors.

File: tests/utils/index.js
Comment (dianabarsan) [inline tests/utils/index.js:1225]: This is so great! Thanks so much for figuring out this workaround

**Review by dianabarsan — COMMENTED** [review]
Great stuff! Bummer about the dependencies. I don't really understand what workspaces are and how they're actually supposed to work (and found no comprehensive guide online). Thanks for trying things with me.

File: api/package-lock.json
Comment (lorerod) [reply api/package-lock.json:283]: It is not part of this work. I will reverting this change.

**Review by dianabarsan — APPROVED** [review]
Finally :D

**Pull request merged**
